Lib9c/Delegation/RewardBase.cs

```diff
@@ -0,0 +1,188 @@
+                    .Select(r => r.Value.Serialize())))
+                .Add(SigFig);
+
+        IValue IBencodable.Bencoded => Bencoded;
+
+        public RewardBase AddRewards(IEnumerable<FungibleAssetValue> rewards)
+            => rewards.Aggregate(this, (accum, next) => AddReward(accum, next));
+
+        public RewardBase AddReward(FungibleAssetValue reward)
+            => AddReward(this, reward);
+
+        public RewardBase UpdateTotalShares(BigInteger totalShares)
+            => UpdateTotalShares(this, totalShares);
+
+        public static RewardBase AddReward(RewardBase rewardBase, FungibleAssetValue reward)
+            => new RewardBase(
+                rewardBase.Address,
+                rewardBase.TotalShares,
+                rewardBase.RewardPortion.TryGetValue(reward.Currency, out var portion)
+                    ? rewardBase.RewardPortion.SetItem(
+                        reward.Currency,
+                        portion + (reward * rewardBase.SigFig).DivRem(rewardBase.TotalShares).Quotient)
+                    : throw new ArgumentException($"Invalid reward currency: {reward.Currency}"),
+                rewardBase.SigFig);
+
+        public static RewardBase UpdateTotalShares(RewardBase rewardBase, BigInteger totalShares)
+        {
+            var newSigFig = Math.Max(rewardBase.SigFig, RecommendedSigFig(totalShares));
+            var multiplier = BigInteger.Pow(10, newSigFig - rewardBase.SigFig);
+            var newPortion = rewardBase.RewardPortion.ToImmutableDictionary(
+                kvp => kvp.Key,
+                kvp => (kvp.Value * multiplier).DivRem(rewardBase.TotalShares).Quotient);
+
+            return new RewardBase(
+                rewardBase.Address,
+                totalShares,
+                newPortion,
+                newSigFig);
+        }
+
+        public static int RecommendedSigFig(BigInteger totalShares)
+            => (int)Math.Floor(BigInteger.Log10(totalShares)) + Margin;
+
+        public ImmutableSortedDictionary<Currency, FungibleAssetValue> RewardsDuringPeriod(BigInteger share)
+            => RewardPortion.Keys.Select(k => RewardsDuringPeriod(share, k))
+                .ToImmutableSortedDictionary(f => f.Currency, f => f, _currencyComparer);
+
+        public FungibleAssetValue RewardsDuringPeriod(BigInteger share, Currency currency)
+            => RewardPortion.TryGetValue(currency, out var portion)
+                ? (portion * share).DivRem(SigFig).Quotient
+                : throw new ArgumentException($"Invalid reward currency: {currency}");
+
+        public override bool Equals(object? obj)
+            => obj is RewardBase other && Equals(other);
+
+        public bool Equals(RewardBase? other)
+            => ReferenceEquals(this, other)
+            || (other is RewardBase rewardBase
+            && Address == rewardBase.Address
+            && TotalShares == rewardBase.TotalShares
+            && RewardPortion.Equals(rewardBase.RewardPortion)
```

Comment:
The code below indicates that the two objects are not the same.
Therefore, the Equals method will likely not work correctly.

```csharp

        var d1 = ImmutableDictionary<string, int>.Empty.Add("a", 1);
        var d2 = ImmutableDictionary<string, int>.Empty.Add("a", 1);

        var t = d1.Equals(d2);
```